pynxtools/dataconverter/readers/multi/reader.py

```diff
@@ -0,0 +1,208 @@
+    """
+    A reader that takes a mapping json file and a data file/object to return a template.
+    """
+
+    # Whitelist for the NXDLs that the reader supports and can process
+    supported_nxdls: List[str] = []
+    extensions: Dict[str, Callable[[Any], dict]] = {
+        ".yaml": parse_yml,
+        ".yml": parse_yml,
+    }
+    kwargs: Optional[Dict[str, Any]] = None
+
+    def __init__(self):
+        self.extensions[".json"] = lambda fn: parse_json_config(
+            fn,
+            ParseJsonCallbacks(
+                attrs_callback=self.get_attr,
+                data_callback=self.get_data,
+                dims=self.get_data_dims,
+            ),
+        )
+
+    def setup_template(self) -> Dict[str, Any]:
+        """
+        Setups the initial data in the template.
+        This may be used to set fixed information, e.g., about the reader.
+        """
+        return {}
+
+    # pylint: disable=unused-argument
+    def handle_objects(self, objects: Tuple[Any]) -> Dict[str, Any]:
+        """
+        Handles the objects passed into the reader.
+        """
+        return {}
+
+    def get_attr(self, path: str) -> Any:
+        """
+        Returns an attributes from the given path.
+        """
+        return {}
+
+    def get_data(self, path: str) -> Any:
+        """
+        Returns data from the given path.
+        """
+        return {}
+
+    def get_data_dims(self, path: str) -> List[str]:
+        """
+        Returns the dimensions of the data from the given path.
+        """
+        return []
+
+    def read(
+        self,
+        template: dict = None,
+        file_paths: Tuple[str] = None,
+        objects: Tuple[Any] = None,
+        **kwargs,
+    ) -> dict:
+        """
+        Reads data from multiple files and passes them to the appropriate functions
+        in the extensions dict.
+        """
+        template = Template()
+        self.kwargs = kwargs
+
+        def sort_keys(filename: str) -> str:
+            """
+            Makes sure to read json and yaml files last
+            """
+            ending = os.path.splitext(filename)[1]
+            return ".zzzzz" if ending in (".json", ".yaml", ".yml") else ending
+
+        sorted_paths = sorted(file_paths, key=sort_keys)
+        for file_path in sorted_paths:
+            extension = os.path.splitext(file_path)[1].lower()
+            if extension not in self.extensions:
+                print(
+                    f"WARNING: "
+                    f"File {file_path} has an unsupported extension, ignoring file."
+                )
+                continue
+            if not os.path.exists(file_path):
+                print(f"WARNING: File {file_path} does not exist, ignoring entry.")
+                continue
+
+            template.update(self.extensions.get(extension, lambda _: {})(file_path))
+
+        template.update(self.setup_template())
+        template.update(self.handle_objects(objects))
```

Comment:
I am not sure what you mean by objects here.

Reply:
You can pass a python object to the read function. We use this in the mpes reader to directly inject an xarray for the data. Since the object doesn't have a file name it has its own function.

Reply:
Could you please point me to where you are cunrrently doing that in the mpes reader? I am not sure I fully understand it yet, but might be useful for me too.

Reply:
Yes, we read the object here: https://github.com/FAIRmat-NFDI/pynxtools-mpes/blob/c1ecb70fe80c8f44cd4d6bc01bf5c2429fa98462/pynxtools_mpes/reader.py#L250 (it get's just passed down into this function from the read function itself)

In this example we pass an xarray to the convert function (in the second to last section called "Create a NeXus file from a xarray"): https://gitlab.mpcdf.mpg.de/nomad-lab/north/mpes/-/blob/main/example/E2%20Binning%20of%20WSe2.ipynb?ref_type=heads

Reply:
And actually we mistreat it a bit as we just pass an xarray object and not a tuple with one object (as the signature indicates). Maybe we can also discuss whether we wan't to allow type `Any` for the objects so it's completely up to the reader what to expect.

We have, however, also a mechanism for passing `**kwargs` down to the read function. This might even be a better option for readers to specialize their input to certain objects.